src/js/components/DataTable/Header.js

```diff
@@ -195,35 +195,28 @@ const Header = forwardRef(
         (selected && enabled.filter((v) => selected.includes(v))) ||
         primaryValues;
 
-      const allSelected = groupBy?.select
-        ? groupBy.select[''] === 'all'
-        : enabledSelected.length === enabled.length;
-
-      if (allSelected) {
-        // if any are disabled and selected, leave those, otherwise clear
-        nextSelected = disabled
-          ? primaryValues.filter(
-              (v) => disabled.includes(v) && selected.includes(v),
-            )
-          : [];
-        nextGroupSelected[''] = 'none';
+      // if all enabled are already selected, remove them from selected;
+      // otherwise add them.
+      if (
+        selected.filter((s) => enabledSelected.includes(s)).length ===
+        enabled.length
+      ) {
+        enabledSelected.forEach((p) => {
+          const index = nextSelected.indexOf(p);
+          if (index >= 0) {
+            nextSelected.splice(index, 1);
+          }
+        });
       } else {
-        // if some or none are selected, select all enabled plus all disabled
-        // that are already selected
-        nextSelected = disabled
-          ? primaryValues.filter(
-              (v) => !disabled.includes(v) || selected.includes(v),
-            )
-          : primaryValues;
-        nextGroupSelected[''] = 'all';
-        groupBy?.expandable?.forEach((key) => {
-          nextGroupSelected[key] = 'all';
+        enabled.forEach((p) => {
+          if (!nextSelected.includes(p)) {
+            nextSelected.push(p);
+          }
         });
       }
-      if (groupBy?.onSelect) {
```

Comment:
Is it backwards compatible to remove this condition? Feels like we might be losing support for something. Doesn't look like we currently have a unit test for `groupBy.onSelect` so would be worth adding to confirm we're staying backwards compatible here.

Reply:
Agree that a unit test would be beneficial here.

Is this story reasonable to confirm desired behavior? https://5d9774839a6eff00203f5cbf-wkumjyibuc.chromatic.com/?path=/story/visualizations-datatable-grouped-and-onselect--grouped-on-select-data-table

Reply:
This story doesn't quite look like it's using `groupBy.onSelect`. It just seems to be using `groupBy` and separately `onSelect` prop but docs say you can have an `onSelect` handler in your groupBy object.

I'll add a unit test and we can check if it's breaking on this branch.

Reply:
addressed